packages/contracts/src/EmailAuth.sol

```diff
@@ -0,0 +1,182 @@
+        require(msg.sender == owner, "only owner can update verifier");
+        require(_verifierAddr != address(0), "invalid verifier address");
+        verifier = Verifier(_verifierAddr);
+    }
+
+    function updateDKIMRegistry(address _dkimRegistryAddr) public {
+        require(msg.sender == owner, "only owner can update dkim registry");
+        require(_dkimRegistryAddr != address(0), "invalid dkim registry address");
+        dkim = ECDSAOwnedDKIMRegistry(_dkimRegistryAddr);
+    }
+
+    function insertSubjectTemplate(uint _templateId, string[] memory _subjectTemplate) public {
+        require(msg.sender == owner, "only owner can insert subject template");
+        require(_subjectTemplate.length > 0, "subject template is empty");
+        require(subjectTemplates[_templateId].length == 0, "template id already exists");
+        subjectTemplates[_templateId] = _subjectTemplate;
+    }
+
+    function updateSubjectTemplate(uint _templateId, string[] memory _subjectTemplate) public {
+        require(msg.sender == owner, "only owner can update subject template");
+        require(_subjectTemplate.length > 0, "subject template is empty");
+        require(subjectTemplates[_templateId].length > 0, "template id not exists");
+        subjectTemplates[_templateId] = _subjectTemplate;
+    }
+
+    function deleteSubjectTemplate(uint _templateId) public {
+        require(msg.sender == owner, "only owner can delete subject template");
+        require(subjectTemplates[_templateId].length > 0, "template id not exists");
+        delete subjectTemplates[_templateId];
+    }
+
+    function computeMsgHash(bytes32 _accountSalt, bool _isCodeExist, uint _templateId, bytes[] memory _subjectParams) public pure returns (bytes32) {
+        return keccak256(abi.encode(_accountSalt, _isCodeExist, _templateId, _subjectParams));
+    }
+
+    function authEmail(EmailAuthMsg memory emailAuthMsg) public returns (bytes32) {
+        require(msg.sender == owner, "only owner can auth email");
+        string[] memory template = subjectTemplates[emailAuthMsg.templateId];
+        require(template.length > 0, "template id not exists");
+        require(
+            dkim.isDKIMPublicKeyHashValid(
+                emailAuthMsg.proof.domainName, 
+                emailAuthMsg.proof.publicKeyHash
+            ) == true, 
+            "invalid dkim public key hash");
+        require(usedNullifiers[emailAuthMsg.proof.emailNullifier] == false, "email nullifier already used");
+        usedNullifiers[emailAuthMsg.proof.emailNullifier] = true;
+        require(accountSalt == emailAuthMsg.proof.accountSalt, "invalid account salt");
+        require(emailAuthMsg.proof.timestamp > 0 && emailAuthMsg.proof.timestamp > lastTimestamp, "invalid timestamp");
+        lastTimestamp = emailAuthMsg.proof.timestamp;
+        
+        // Construct an expectedSubject from template and the values of emailAuthMsg.subjectParams.
```

Comment:
Is this string comparison different from what was written in the spec? Maybe my imagined intention is wrong.
Also, I would like a reference example of what title to use for the unit test.

Reply:
I meant that if and only if `emailAuthMsg.proof.timestamp > 0`, `emailAuthMsg.proof.timestamp > lastTimestamp` should be checked.
In other words, this check is defined as `emailAuthMsg.proof.timestamp == 0 || emailAuthMsg.proof.timestamp > lastTimestamp`.